File: clients/csharp/LightKV/LightKVClient.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Net.Sockets;
using System.Text;
using System.Threading.Tasks;

namespace LightKV
{
    /// <summary>
    /// LightKV C# 客户端 SDK
    ///
    /// 通过 TCP 使用 Redis RESP 协议与 LightKV 服务通信。
    /// 支持连接池、自动重连、Pipeline 批量操作。
    ///
    /// 用法：
    ///   using var client = new LightKVClient("127.0.0.1", 6379);
    ///   client.Set("key", "value");
    ///   string? val = client.Get("key");
    /// </summary>
    public class LightKVClient : IDisposable
    {
        private readonly string _host;
        private readonly int _port;
        private readonly int _timeoutMs;
        private readonly int _retryCount;

        private TcpClient? _tcpClient;
        private NetworkStream? _stream;
        private readonly object _lock = new();

        // Pipeline support
        private List<string>? _pipelineBuf;

        public LightKVClient(string host, int port, int timeoutMs = 3000, int retryCount = 3)
        {
            _host = host;
            _port = port;
            _timeoutMs = timeoutMs;
            _retryCount = retryCount;
        }

        // ─── 连接管理 ───

        public bool Connect()
        {
            lock (_lock)
            {
                try
                {
                    _tcpClient = new TcpClient();
                    var task = _tcpClient.ConnectAsync(_host, _port);
                    task.Wait(_timeoutMs);
                    if (!_tcpClient.Connected) return false;
                    _stream = _tcpClient.GetStream();
                    _stream.ReadTimeout = _timeoutMs;
                    _stream.WriteTimeout = _timeoutMs;
                    return true;
                }
                catch
                {
                    return false;
                }
            }
        }

        public void Dispose()
        {
            lock (_lock)
            {
                _stream?.Close();
                _tcpClient?.Close();
                _stream = null;
                _tcpClient = null;
            }
        }

        public bool IsConnected
        {
            get { lock (_lock) return _tcpClient?.Connected ?? false; }
        }

        // ─── 核心 KV 操作 ───

        public bool Set(string key, string value)
        {
            var resp = SendCommand("SET", key, value);
            return resp == "+OK";
        }

        public string? Get(string key)
        {
            return SendCommand("GET", key) switch
            {
                null or "$-1" => null,
                var s when s.StartsWith("$") => ParseBulkString(s),
                var s when s.StartsWith("+") => s[1..],
                _ => null
            };
        }

        public bool Delete(string key)
        {
            var resp = SendCommand("DEL", key);
            return resp != null && resp.StartsWith(":1");
        }

        public bool Exists(string key)
        {
            var resp = SendCommand("EXISTS", key);
            return resp != null && resp.StartsWith(":1");
        }

        public long Incr(string key)
        {
            var resp = SendCommand("INCR", key);
            return ParseInteger(resp);
        }

        public long Decr(string key)
        {
            var resp = SendCommand("DECR", key);
            return ParseInteger(resp);
        }

        public bool Expire(string key, long seconds)
        {
            var resp = SendCommand("EXPIRE", key, seconds.ToString());
            return resp != null && resp.StartsWith(":1");
        }

        public long Ttl(string key)
        {
            var resp = SendCommand("TTL", key);
            return ParseInteger(resp);
        }

        // ─── Hash 操作 ───

        public bool HSet(string key, string field, string value)
        {
            var resp = SendCommand("HSET", key, field, value);
            return resp != null && resp.StartsWith(":1");
        }

        public string? HGet(string key, string field)
        {
            var resp = SendCommand("HGET", key, field);
            return resp switch
            {
                null or "$-1" => null,
                var s when s.StartsWith("$") => ParseBulkString(s),
                _ => null
            };
        }

        public Dictionary<string, string> HGetAll(string key)
        {
            var result = new Dictionary<string, string>();
            var resp = SendCommand("HGETALL", key);
            if (resp == null) return result;
            var items = ParseArray(resp);
            for (int i = 0; i + 1 < items.Count; i += 2)
            {
                result[items[i]] = items[i + 1];
            }
            return result;
        }

        // ─── List 操作 ───

        public long LPush(string key, params string[] values)
        {
            var args = new List<string> { "LPUSH", key };
            args.AddRange(values);
            var resp = SendCommand(args.ToArray());
            return ParseInteger(resp);
        }

        public string? LPop(string key)
        {
            var resp = SendCommand("LPOP", key);
            return resp switch
            {
                null or "$-1" => null,
                var s when s.StartsWith("$") => ParseBulkString(s),
                _ => null
            };
        }

        public List<string> LRange(string key, long start, long stop)
        {
            var resp = SendCommand("LRANGE", key, start.ToString(), stop.ToString());
            return resp != null ? ParseArray(resp) : new List<string>();
        }

        // ─── Set 操作 ───

        public long SAdd(string key, params string[] members)
        {
            var args = new List<string> { "SADD", key };
            args.AddRange(members);
            var resp = SendCommand(args.ToArray());
            return ParseInteger(resp);
        }

        public HashSet<string> SMembers(string key)
        {
            var resp = SendCommand("SMEMBERS", key);
            return resp != null ? new HashSet<string>(ParseArray(resp)) : new HashSet<string>();
        }

        // ─── ZSet 操作 ───

        public long ZAdd(string key, double score, string member)
        {
            var resp = SendCommand("ZADD", key, score.ToString("F6"), member);
            return ParseInteger(resp);
        }

        public List<string> ZRange(string key, long start, long stop)
        {
            var resp = SendCommand("ZRANGE", key, start.ToString(), stop.ToString());
            return resp != null ? ParseArray(resp) : new List<string>();
        }

        // ─── Pipeline 支持 ───

        public void Pipeline() => _pipelineBuf = new List<string>();

        public void Queue(params string[] args)
        {
            _pipelineBuf?.Add(BuildResp(args));
        }

        public List<string?> ExecPipeline()
        {
            if (_pipelineBuf == null || _pipelineBuf.Count == 0)
                return new List<string?>();

            lock (_lock)
            {
                try
                {
                    EnsureConnected();
                    var sb = new StringBuilder();
                    foreach (var cmd in _pipelineBuf) sb.Append(cmd);
                    var data = Encoding.UTF8.GetBytes(sb.ToString());
                    _stream!.Write(data, 0, data.Length);

                    var results = new List<string?>();
                    foreach (var _ in _pipelineBuf)
                    {
                        results.Add(ReadLine());
                    }
                    _pipelineBuf = null;
                    return results;
                }
                catch
                {
                    _pipelineBuf = null;
                    return new List<string?>();
                }
            }
        }

        // ─── 通用命令 ───

        public bool Ping()
        {
            var resp = SendCommand("PING");
            return resp == "+PONG";
        }

        // ─── 内部方法 ───

        private string? SendCommand(params string[] args)
        {
            for (int attempt = 0; attempt < _retryCount; attempt++)
            {
                try
                {
                    lock (_lock)
                    {
                        EnsureConnected();
                        var cmd = BuildResp(args);
                        var data = Encoding.UTF8.GetBytes(cmd);
                        _stream!.Write(data, 0, data.Length);
                        return ReadLine();
                    }
                }
                catch
                {
                    Dispose();
                    if (attempt < _retryCount - 1)
                        Task.Delay(100 * (attempt + 1)).Wait();
                }
            }
            return null;
        }

        private void EnsureConnected()
        {
            if (_tcpClient == null || !_tcpClient.Connected)
            {
                Dispose();
                Connect();
            }
        }

        private string? ReadLine()
        {
            if (_stream == null) return null;
            var sb = new StringBuilder();
            int b;
            while ((b = _stream.ReadByte()) != -1)
            {
                if (b == '\r')
                {
                    _stream.ReadByte(); // consume \n
                    break;
                }
                sb.Append((char)b);
            }
            return sb.Length > 0 ? sb.ToString() : null;
        }

        private static string BuildResp(string[] args)
        {
            var sb = new StringBuilder();
            sb.Append('*').Append(args.Length).Append("\r\n");
            foreach (var arg in args)
            {
                var bytes = Encoding.UTF8.GetByteCount(arg);
                sb.Append('$').Append(bytes).Append("\r\n");
                sb.Append(arg).Append("\r\n");
            }
            return sb.ToString();
        }

        private string? ParseBulkString(string resp)
        {
            if (resp.StartsWith("$-1")) return null;
            if (!resp.StartsWith("$")) return resp;

            var lenStr = resp[1..];
            if (!int.TryParse(lenStr, out int len) || len < 0) return null;

            var buf = new byte[len];
            _stream?.Read(buf, 0, len);
            // Read trailing \r\n
            _stream?.ReadByte();
            _stream?.ReadByte();
            return Encoding.UTF8.GetString(buf);
        }

        private static long ParseInteger(string? resp)
        {
            if (resp == null || !resp.StartsWith(":")) return 0;
            long.TryParse(resp[1..], out long val);
            return val;
        }

        private List<string> ParseArray(string resp)
        {
            var result = new List<string>();
            if (resp.StartsWith("*-1") || resp.StartsWith("*0")) return result;

            if (!resp.StartsWith("*") || !int.TryParse(resp[1..], out int count))
                return result;

            for (int i = 0; i < count; i++)
            {
                var line = ReadLine();
                if (line == null) break;
                if (line.StartsWith("$"))
                {
                    if (!int.TryParse(line[1..], out int len) || len < 0) continue;
                    var buf = new byte[len];
                    _stream?.Read(buf, 0, len);
                    _stream?.ReadByte(); _stream?.ReadByte(); // \r\n
                    result.Add(Encoding.UTF8.GetString(buf));
                }
                else
                {
                    result.Add(line);
                }
            }
            return result;
        }
    }
}

File: clients/java/src/main/java/lightkv/LightKVClient.java
package lightkv;

import java.io.*;
import java.net.*;
import java.util.*;
import java.util.concurrent.*;

/**
 * LightKV Java 客户端 SDK
 *
 * 通过 TCP 使用 Redis RESP 协议与 LightKV 服务通信。
 * 支持连接池、自动重连、Pipeline 批量操作。
 *
 * 用法：
 *   LightKVClient client = new LightKVClient("127.0.0.1", 6379);
 *   client.set("key", "value");
 *   String val = client.get("key").orElse(null);
 *   client.close();
 */
public class LightKVClient implements AutoCloseable {
    private final String host;
    private final int port;
    private final int timeoutMs;
    private final int retryCount;

    private Socket socket;
    private BufferedReader reader;
    private OutputStream writer;
    private String lastError;
    private final Object lock = new Object();

    // Pipeline support
    private List<String> pipelineBuf;

    public LightKVClient(String host, int port) {
        this(host, port, 3000, 3);
    }

    public LightKVClient(String host, int port, int timeoutMs, int retryCount) {
        this.host = host;
        this.port = port;
        this.timeoutMs = timeoutMs;
        this.retryCount = retryCount;
    }

    // ─── 连接管理 ───

    public boolean connect() {
        synchronized (lock) {
            try {
                socket = new Socket();
                socket.connect(new InetSocketAddress(host, port), timeoutMs);
                socket.setSoTimeout(timeoutMs);
                reader = new BufferedReader(new InputStreamReader(socket.getInputStream(), "UTF-8"));
                writer = socket.getOutputStream();
                lastError = null;
                return true;
            } catch (IOException e) {
                lastError = e.getMessage();
                return false;
            }
        }
    }

    public void close() {
        synchronized (lock) {
            try { if (socket != null) socket.close(); } catch (IOException ignored) {}
            socket = null;
            reader = null;
            writer = null;
        }
    }

    public boolean isConnected() {
        synchronized (lock) {
            return socket != null && socket.isConnected() && !socket.isClosed();
        }
    }

    public String getLastError() { return lastError; }

    // ─── 核心 KV 操作 ───

    public boolean set(String key, String value) {
        return sendCommand("SET", key, value).map(r -> r.equals("+OK\r\n")).orElse(false);
    }

    public Optional<String> get(String key) {
        return sendCommand("GET", key).flatMap(this::parseBulkString);
    }

    public boolean delete(String key) {
        return sendCommand("DEL", key).map(r -> r.startsWith(":1")).orElse(false);
    }

    public boolean exists(String key) {
        return sendCommand("EXISTS", key).map(r -> r.startsWith(":1")).orElse(false);
    }

    public long incr(String key) {
        return sendCommand("INCR", key).map(this::parseInteger).orElse(0L);
    }

    public long decr(String key) {
        return sendCommand("DECR", key).map(this::parseInteger).orElse(0L);
    }

    public boolean expire(String key, long seconds) {
        return sendCommand("EXPIRE", key, String.valueOf(seconds))
                .map(r -> r.startsWith(":1")).orElse(false);
    }

    public long ttl(String key) {
        return sendCommand("TTL", key).map(this::parseInteger).orElse(-2L);
    }

    // ─── Hash 操作 ───

    public boolean hset(String key, String field, String value) {
        return sendCommand("HSET", key, field, value).map(r -> r.startsWith(":1")).orElse(false);
    }

    public Optional<String> hget(String key, String field) {
        return sendCommand("HGET", key, field).flatMap(this::parseBulkString);
    }

    public Map<String, String> hgetAll(String key) {
        return sendCommand("HGETALL", key).map(this::parseArray).orElse(Collections.emptyList())
                .stream().collect(HashMap::new, (m, v) -> {
                    // pairs: k1, v1, k2, v2, ...
                }, HashMap::putAll);
    }

    // ─── List 操作 ───

    public long lpush(String key, String... values) {
        List<String> args = new ArrayList<>(Arrays.asList("LPUSH", key));
        args.addAll(Arrays.asList(values));
        return sendCommand(args.toArray(new String[0])).map(this::parseInteger).orElse(0L);
    }

    public Optional<String> lpop(String key) {
        return sendCommand("LPOP", key).flatMap(this::parseBulkString);
    }

    public List<String> lrange(String key, long start, long stop) {
        return sendCommand("LRANGE", key, String.valueOf(start), String.valueOf(stop))
                .map(this::parseArray).orElse(Collections.emptyList());
    }

    // ─── Set 操作 ───

    public long sadd(String key, String... members) {
        List<String> args = new ArrayList<>(Arrays.asList("SADD", key));
        args.addAll(Arrays.asList(members));
        return sendCommand(args.toArray(new String[0])).map(this::parseInteger).orElse(0L);
    }

    public Set<String> smembers(String key) {
        return new HashSet<>(sendCommand("SMEMBERS", key)
                .map(this::parseArray).orElse(Collections.emptyList()));
    }

    // ─── ZSet 操作 ───

    public long zadd(String key, double score, String member) {
        return sendCommand("ZADD", key, String.valueOf(score), member)
                .map(this::parseInteger).orElse(0L);
    }

    public List<String> zrange(String key, long start, long stop) {
        return sendCommand("ZRANGE", key, String.valueOf(start), String.valueOf(stop))
                .map(this::parseArray).orElse(Collections.emptyList());
    }

    // ─── Pipeline 支持 ───

    public void pipeline() {
        pipelineBuf = new ArrayList<>();
    }

    public void queue(String... args) {
        if (pipelineBuf != null) {
            pipelineBuf.add(buildResp(args));
        }
    }

    public List<Optional<String>> execPipeline() {
        if (pipelineBuf == null || pipelineBuf.isEmpty()) return Collections.emptyList();
        synchronized (lock) {
            try {
                StringBuilder sb = new StringBuilder();
                for (String cmd : pipelineBuf) sb.append(cmd);
                writer.write(sb.toString().getBytes("UTF-8"));
                writer.flush();

                List<Optional<String>> results = new ArrayList<>();
                for (int i = 0; i < pipelineBuf.size(); i++) {
                    String line = reader.readLine();
                    results.add(Optional.ofNullable(line));
                }
                pipelineBuf = null;
                return results;
            } catch (IOException e) {
                lastError = e.getMessage();
                pipelineBuf = null;
                return Collections.emptyList();
            }
        }
    }

    // ─── 内部方法 ───

    private Optional<String> sendCommand(String... args) {
        synchronized (lock) {
            for (int attempt = 0; attempt < retryCount; attempt++) {
                try {
                    if (socket == null || !socket.isConnected()) {
                        if (!connect()) continue;
                    }
                    String cmd = buildResp(args);
                    writer.write(cmd.getBytes("UTF-8"));
                    writer.flush();
                    return Optional.of(reader.readLine());
                } catch (IOException e) {
                    lastError = e.getMessage();
                    close();
                    try { Thread.sleep(100L * (attempt + 1)); } catch (InterruptedException ignored) {}
                }
            }
            return Optional.empty();
        }
    }

    private String buildResp(String... args) {
        StringBuilder sb = new StringBuilder();
        sb.append('*').append(args.length).append("\r\n");
        for (String arg : args) {
            byte[] bytes = arg.getBytes(java.nio.charset.StandardCharsets.UTF_8);
            sb.append('$').append(bytes.length).append("\r\n");
            sb.append(arg).append("\r\n");
        }
        return sb.toString();
    }

    private Optional<String> parseBulkString(String resp) {
        if (resp == null || resp.startsWith("$-1")) return Optional.empty();
        if (resp.startsWith("+")) return Optional.of(resp.substring(1));
        if (resp.startsWith("$")) {
            // Read the actual data following the length
            try {
                int len = Integer.parseInt(resp.substring(1));
                if (len < 0) return Optional.empty();
                char[] data = new char[len];
                int read = reader.read(data, 0, len);
                reader.readLine(); // consume trailing \r\n
                return Optional.of(new String(data, 0, read));
            } catch (IOException e) {
                lastError = e.getMessage();
                return Optional.empty();
            }
        }
        return Optional.of(resp);
    }

    private long parseInteger(String resp) {
        if (resp == null || resp.isEmpty()) return 0;
        try {
            if (resp.startsWith(":")) return Long.parseLong(resp.substring(1));
        } catch (NumberFormatException ignored) {}
        return 0;
    }

    private List<String> parseArray(String resp) {
        List<String> result = new ArrayList<>();
        if (resp == null || resp.isEmpty() || resp.startsWith("*-1") || resp.startsWith("*0")) return result;
        try {
            int count = Integer.parseInt(resp.substring(1));
            for (int i = 0; i < count; i++) {
                String line = reader.readLine();
                if (line != null && line.startsWith("$")) {
                    int len = Integer.parseInt(line.substring(1));
                    if (len >= 0) {
                        char[] data = new char[len];
                        reader.read(data, 0, len);
                        result.add(new String(data));
                        reader.readLine(); // consume trailing \r\n
                    }
                }
            }
        } catch (IOException | NumberFormatException e) {
            lastError = e.getMessage();
        }
        return result;
    }
}
